### PR TITLE
Part 2

### PR DESCRIPTION
**Pull request by ruancomelli**
## Summary by Sourcery

This pull request introduces a new 'Class Distribution Tool' page for managing class distributions, refactors the 'create_classes' function for better clarity, and updates multiple JSON files to standardize department references.

- **New Features**:
    - Introduced a new 'Class Distribution Tool' page with a form for creating and managing class distributions, including fields for group title, year, stage, faculty, department, time, and semester.
- **Enhancements**:
    - Refactored the 'create_classes' function to accept individual parameters instead of a document object, improving clarity and maintainability.
    - Updated various JSON files to replace 'Faculty Department' with 'Department' for consistency and accuracy in department references.
    - Modified the 'Graduation Year' field in the 'Student' doctype to be a selectable list of years instead of a date field.

### CODE REVIEW

**Comment by ruancomelli** [comment]
This is an experiment review for experiment `improve_review_guide`
Run ID: `improve_review_guide/run_2024-08-12T01-25-21_v1-21-0-75-g76dae36a2`
The benchmark review for this pull request can be found at https://github.com/sourcery-ai-experiments/kalima/pull/17.
This pull request was cloned from `https://github.com/sourcery-ai-experiments/kalima/pull/3`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_

<details>
<summary>Experiment configuration</summary>

```yaml
review_config:
  # User configuration for the review
  # - benchmark - use the user config from the benchmark reviews
  # - <value> - use the value directly
  user_review_config:
    enable_ai_review: true
    enable_rule_comments: false

    enable_complexity_comments: false
    enable_security_comments: false
    enable_tests_comments: false
    enable_comment_suggestions: false

    enable_claude_comments: true
    enable_functionality_review: false

    enable_pull_request_summary: false
    enable_review_guide: true

    enable_approvals: false

  ai_review_config:
    # The model responses to use for the experiment
    # - benchmark - use the model responses from the benchmark reviews
    # - llm - call the language model to generate responses
    model_responses:
      comments_model: benchmark
      comment_area_model: benchmark
      comment_validation_model: benchmark
      comment_suggestion_model: benchmark
      complexity_model: benchmark
      docstrings_model: benchmark
      functionality_model: benchmark
      overall_comments_model: benchmark
      security_model: benchmark
      tests_model: benchmark
      pull_request_summary_model: benchmark
      review_guide_model: llm

# The pull request dataset to run the experiment on
pull_request_dataset:
  # open-source PRs
- https://github.com/sourcery-ai-experiments/gpt-crawler/pull/1
- https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/1
- https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/2
- https://github.com/sourcery-ai-experiments/iptux/pull/5
- https://github.com/sourcery-ai-experiments/iptux/pull/6
- https://github.com/sourcery-ai-experiments/kalima/pull/3
- https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/1
- https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/2
- https://github.com/sourcery-ai-experiments/Trenova/pull/11
  # Sourcery PRs
- https://github.com/sourcery-ai/core/pull/4055
- https://github.com/sourcery-ai/core/pull/4263
- https://github.com/sourcery-ai/core/pull/4352
- https://github.com/sourcery-ai/core/pull/4364
- https://github.com/sourcery-ai/core/pull/4366
- https://github.com/sourcery-ai/core/pull/4390
- https://github.com/sourcery-ai/core/pull/4402
- https://github.com/sourcery-ai/core/pull/4445
- https://github.com/sourcery-ai/core/pull/4462
- https://github.com/sourcery-ai/core/pull/4463

review_comment_labels: []

# Benchmark reviews generated by running
#   python -m scripts.experiment benchmark <experiment_name>
benchmark_reviews:
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4463
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/236
- dataset_pull_request: https://github.com/sourcery-ai-experiments/gpt-crawler/pull/1
  review_pull_request: https://github.com/sourcery-ai-experiments/gpt-crawler/pull/10
- dataset_pull_request: https://github.com/sourcery-ai-experiments/iptux/pull/5
  review_pull_request: https://github.com/sourcery-ai-experiments/iptux/pull/32
- dataset_pull_request: https://github.com/sourcery-ai-experiments/iptux/pull/6
  review_pull_request: https://github.com/sourcery-ai-experiments/iptux/pull/33
- dataset_pull_request: https://github.com/sourcery-ai-experiments/kalima/pull/3
  review_pull_request: https://github.com/sourcery-ai-experiments/kalima/pull/17
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4055
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/258
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4263
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/259
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4352
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/260
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4364
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/261
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4366
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/262
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4390
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/263
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4402
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/264
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4445
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/265
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4462
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/266
- dataset_pull_request: https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/1
  review_pull_request: https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/23
- dataset_pull_request: https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/24
- dataset_pull_request: https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/1
  review_pull_request: https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/19
- dataset_pull_request: https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/20
- dataset_pull_request: https://github.com/sourcery-ai-experiments/Trenova/pull/11
  review_pull_request: https://github.com/sourcery-ai-experiments/Trenova/pull/19

```

</details>

**Comment by SourceryAI** [comment]
## Reviewer's Guide by Sourcery

This pull request introduces a new 'Class Distribution Tool' page and makes several updates to existing files, primarily focusing on standardizing department references and refactoring the class creation process. The changes include:

1. A new page 'Class Distribution Tool' has been added, providing a user interface for managing class distributions.
2. The 'create_classes' function in the GroupClass doctype has been refactored to accept individual parameters instead of a document object, improving clarity and maintainability.
3. Multiple JSON files have been updated to replace 'Faculty Department' references with 'Department' for consistency.
4. The 'Graduation Year' field in the 'Student' doctype has been modified to be a selectable list of years instead of a date field.
5. Various other minor updates and refactorings have been made across multiple files to support these changes.

### File-Level Changes

| Files | Changes |
| ----- | ------- |
| `kalima/kalima/doctype/group_class/group_class.py` | Refactored 'create_classes' function to accept individual parameters instead of a document object |
| `kalima/kalima/doctype/student_scores/student_scores.json`<br/>`kalima/kalima/doctype/presented_module/presented_module.json`<br/>`kalima/kalima/doctype/class/class.json`<br/>`kalima/kalima/doctype/applicant_student/applicant_student.json`<br/>`kalima/kalima/doctype/complaints/complaints.json`<br/>`kalima/kalima/doctype/constant/constant.json`<br/>`kalima/kalima/doctype/continuous_exam_result/continuous_exam_result.json`<br/>`kalima/kalima/doctype/department_module/department_module.json`<br/>`kalima/kalima/doctype/exam_departments/exam_departments.json`<br/>`kalima/kalima/doctype/exam_result_entry/exam_result_entry.json`<br/>`kalima/kalima/doctype/exam_students/exam_students.json`<br/>`kalima/kalima/doctype/group_class/group_class.json`<br/>`kalima/kalima/doctype/library/library.json`<br/>`kalima/kalima/doctype/outgoing/outgoing.json`<br/>`kalima/kalima/doctype/preferred_departments/preferred_departments.json`<br/>`kalima/kalima/doctype/presented_department/presented_department.json`<br/>`kalima/kalima/doctype/student_attendance_entry/student_attendance_entry.json`<br/>`kalima/kalima/doctype/student_result_log/student_result_log.json`<br/>`kalima/kalima/doctype/students_fees/students_fees.json`<br/>`kalima/kalima/doctype/transfer/transfer.json` | Updated JSON files to replace 'Faculty Department' with 'Department' for consistency |
| `kalima/kalima/doctype/student/student.json` | Modified 'Graduation Year' field in 'Student' doctype to be a selectable list of years |
| `kalima/kalima/page/class_distribution_t/class_distribution_t.js`<br/>`kalima/kalima/page/class_distribution_t/class_distribution_t.json` | Added new 'Class Distribution Tool' page with form for creating and managing class distributions |
| `kalima/kalima/custom/department.json` | Updated 'Department' doctype with new custom fields and property setters |

---

<details>
<summary>Tips</summary>

- Trigger a new Sourcery review by commenting `@sourcery-ai review` on the pull request.
- Continue your discussion with Sourcery by replying directly to review comments.
- You can change your review settings at any time by accessing your [dashboard](https://app.sourcery.ai):
    - Enable or disable the Sourcery-generated pull request summary or reviewer's guide;
    - Change the review language;
- You can always [contact us](mailto:support@sourcery.ai) if you have any questions or feedback.

</details>

File: kalima/kalima/doctype/group_class/group_class.py [hunk 1]
```diff
@@ -60,17 +62,21 @@ def fetch_students(selected_modules, department):
...
-
-
 @frappe.whitelist()
-def create_classes(group_class_doc,group_class_modules,students):
+def create_classes(group_title,year,stage,semester,department,group_class_modules,students):
     group_class_modules = json.loads(str(group_class_modules))
     for module in group_class_modules:
         # Convert the group_class_doc to a dictionary
-        group_class_doc = json.loads(str(group_class_doc))
+        # group_class_doc = json.loads(str(group_class_doc))
```
Comment (SourceryAI) [inline kalima/kalima/doctype/group_class/group_class.py:70]: **suggestion:** Remove commented-out code.

The commented-out code should be removed if it's not going to be used. It clutters the codebase and can be confusing.

File: kalima/kalima/doctype/group_class/group_class.js [hunk 2]
```diff
@@ -16,18 +16,6 @@ frappe.ui.form.on("Group Class", {
...
-        // const selected_modules = [];
-        // const department = frm.doc.department;
-
-        // // Get selected modules
-        // frm.fields_dict.presented_modules.$wrapper.find('input[type="checkbox"]:checked').each(function () {
-        //     selected_modules.push($(this).val());
-        // });
-
-        // if (selected_modules.length === 0 || !department) {
-        //     return;
```
Comment (SourceryAI) [inline kalima/kalima/doctype/group_class/group_class.js:28]: **suggestion:** Remove commented-out code in `after_save` function.

The commented-out code in the `after_save` function should be removed. It's not serving any purpose.

File: kalima/kalima/page/class_managment/class_managment.js [hunk 3]
```diff
@@ -465,7 +465,8 @@ function createFormDialogNew(templateName) {
                     fields: [
                         { fieldname: 'student_code', fieldtype: 'Data', in_list_view: 1, label: 'Student Code' },
                         { fieldname: 'student_name', fieldtype: 'Link', options: 'Student', in_list_view: 1, label: 'Student Name' },
-                        { fieldname: 'department', fieldtype: 'Link', options: 'Faculty Department', label: 'Department' },
+                        // { fieldname: 'department', fieldtype: 'Link', options: 'Faculty Department', label: 'Department' },
+                        { fieldname: 'department', fieldtype: 'Link', options: 'Department', label: 'Department' },
                         {
```
Comment (SourceryAI) [inline kalima/kalima/page/class_managment/class_managment.js:469]: **suggestion:** Remove commented-out code for `department` field.

The commented-out code for the `department` field should be removed. It makes the code harder to read.

File: kalima/kalima/page/class_distribution_t/class_distribution_t.js [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+        body: page.body
+    });
+
+    form.make();
+
+    let studentListContainer = $('<div id="student-list"></div>').appendTo(page.body);
+    let moduleListContainer = $('<div id="module-list"></div>').appendTo(page.body);
+
+    // Function to fetch and display students
+    function fetch_students() {
```
Comment (SourceryAI) [inline kalima/kalima/page/class_distribution_t/class_distribution_t.js:124]: **issue:** Remove redundant `fetch_students` function definition.

The `fetch_students` function is defined twice. This is redundant and could lead to confusion. We should remove one of them.

**Review by SourceryAI — COMMENTED** [review]
Hey @ruancomelli - I've reviewed your changes - here's some feedback:

**Overall Comments**:
- Remove commented-out code to keep the codebase clean and maintainable.
- Consider adding unit tests for the new 'Class Distribution Tool' page to ensure its functionality.
- Consider generating the 'Graduation Year' list dynamically based on the current year to avoid maintenance issues.
- Remove commented-out code to keep the codebase clean and maintainable.
- Consider adding unit tests for the new 'Class Distribution Tool' page to ensure its functionality.
- Consider generating the 'Graduation Year' list dynamically based on the current year to avoid maintenance issues.

<details>
<summary>Here's what I looked at during the review</summary>

- 🟡 **General issues**: 4 issues found
- 🟢 **Documentation**: all looks good
</details>

***

[LangSmith trace](https://smith.langchain.com/o/7d8518ad-6dd3-5c7b-96d2-757769296335/projects/p/88560225-5072-48bc-967c-dc54da53f265/r/7ea9bcd7-5fa0-4563-acf3-d86a3491a0a9?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment to tell me if it was helpful.
</sub>

**Pull request closed without merging**